### DISCUSSION TITLE
Remove "Preview" watermark/ image from auto generated PDF File

### DISCUSSION THREAD

**Discussion by scaffidic** [pymupdf/PyMuPDF#3699] [Looking for help]
Hi everyone,
I am trying to create a python script that will iterate over every page of an insurance policy to remove the Preview watermark. I have the green light from my company to work on this. The issues I am running into are:

1. The "Preview" watermark is sometimes a textbox and sometimes a image.
    I tried running `get_page_images` but it returns an empty list. I manually added an image and reran the code and it was able to find 
    the image properties perfectly
2. For the textboxes, I've tried to change the text to an empty string and it is replaced by a white box the size of the watermark.
3. I've tried writing logic that only removes a bbox with the same dimensions/ coordinates of the preview watermark and that also 
    creates a white box on top.      
4. I tried to retrieve the watermark metadata to target the specific box with the Preview but, again it returns an empty list.

A potential solution/ workaround could be creating a script that manually opens adobe acrobat and repeats the mouse click and button press to delete the preview watermark on every page unless someone has a solution. I really appreciate the help on this.

```
import fitz  # PyMuPDF
import tkinter as tk
from tkinter import filedialog
import os
[GitHub Example.pdf](https://github.com/user-attachments/files/16288392/GitHub.Example.pdf)


def select_pdf_file():
    root = tk.Tk()
    root.withdraw()  
    file_path = filedialog.askopenfilename(filetypes=[("PDF files", "*.pdf")])
    return file_path

def process_pages(doc, search_text, replace_text):
    for page in doc:
        found = page.search_for(search_text)
        for item in found:
            page.add_redact_annot(item, '')
            page.apply_redactions()
            page.insert_text(item.bl - (0,3), replace_text, fontsize=6)

def main():
    pdf_file_path = select_pdf_file()
    if not pdf_file_path:
        print("No file selected. Exiting.")
        return

    base_name, ext = os.path.splitext(pdf_file_path)

    doc = fitz.open(pdf_file_path)

    for xref in range(1, doc.xref_length()):  # inspect all PDF objects
        try:
            source = doc.xref_object(xref)  # PDF object source
            if not "/Watermark" in source or not "/Form" in source:
                continue  # not a watermark object
        except Exception:  # PDF contains invalid xref numbers
        	continue
        doc.update_object(xref, "<<>>")  # replace by empty dictionary

    # Process for Agent Copy
    process_pages(doc, "ADVANCE COPY", "Agent Copy")
    agent_copy_path = f"{base_name} - Agent Copy{ext}"
    doc.save(agent_copy_path)

    # Process for Insured Copy
    process_pages(doc, "Agent Copy", "Insured Copy")
    insured_copy_path = f"{base_name} - Insured Copy{ext}"
    doc.delete_page(0)
    doc.save(insured_copy_path)

    print(f"Agent Copy saved as: {agent_copy_path}")
    print(f"Insured Copy saved as: {insured_copy_path}")

main()
```

**Comment by scaffidic** [comment]
[GitHub Example.pdf](https://github.com/user-attachments/files/16288409/GitHub.Example.pdf)

**Comment by JorjMcKie** [comment]
I had a look at that example file and did not see how this special way of "watermarking" has been implemented.
As I mentioned in many similar cases: the alternatives of reaching this effect are multiple.
Here definitely no text insertion was used, neither is it the PDF standard way of watermarking or a dedicated annotation type.
Probably a set of drawing instructions repeated in the appearance stream of every page.
If this sequence of instructions could be found, their deletion should be possible. Searching for communalities is too tedious here.

**Reply by scaffidic** [reply]
Hi Jorj,
Can you elaborate on what you mean by set of drawing instructions?

**Reply by JorjMcKie** [reply]
These are instructions as PyMuPDF can generate them: draw_rect, draw_circle, draw_line, ... But obviously in the syntax of PDF's mini-language used in the appearance stream (`/Contents` object). For example, a rectangle with red border and yellow fill color and a circle of radius 2 centered at some point with blue border and no color might be represented by the following instructions in the page's appearance source:
```
q
100 642 100 100 re
h
1 0 0 RG 1 1 0 rg B
Q
 
q
298 542 m
298 540.8954 298.89543 540 300 540 c
301.10459 540 302 540.8954 302 542 c
302 543.1046 301.10459 544 300 544 c
298.89543 544 298 543.1046 298 542 c
h
0 0 1 RG S
Q
```
Fat chance that you do not understand this, and no worries: I am not going to explain it. Please consult the PDF specification manuals in case of interest.
